src/Functions/formatRow.cpp

```diff
@@ -0,0 +1,122 @@
+#include <memory>
+#include <Columns/ColumnString.h>
+#include <DataStreams/materializeBlock.h>
+#include <DataTypes/DataTypeString.h>
+#include <Formats/FormatFactory.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunctionImpl.h>
+#include <IO/WriteBufferFromVector.h>
+#include <IO/WriteHelpers.h>
+#include <Processors/Formats/IOutputFormat.h>
+#include <ext/map.h>
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int NOT_IMPLEMENTED;
+    extern const int UNKNOWN_FORMAT;
+}
+
+
+/** formatRow(<format>, x, y, ...) is a function that allows you to use RowOutputFormat over
+  * several columns to generate a string per row, such as CSV, TSV, JSONEachRow, etc.
+  * formatRowNoNewline(...) trims the newline character of each row.
+  */
+
+template <bool no_newline>
+class FunctionFormatRow : public IFunction
+{
+public:
+    static constexpr auto name = no_newline ? "formatRowNoNewline" : "formatRow";
+
+    FunctionFormatRow(const String & format_name_, const Context & context_) : format_name(format_name_), context(context_)
+    {
+        if (!FormatFactory::instance().getAllFormats().count(format_name))
+            throw Exception("Unknown format " + format_name, ErrorCodes::UNKNOWN_FORMAT);
+    }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 0; }
+    bool useDefaultImplementationForNulls() const override { return false; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+    ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return {0}; }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t input_rows_count) override
+    {
+        auto col_str = ColumnString::create();
+        ColumnString::Chars & vec = col_str->getChars();
+        WriteBufferFromVector buffer(vec);
+        ColumnString::Offsets & offsets = col_str->getOffsets();
+        offsets.resize(input_rows_count);
+        Block arg_block;
+        for (auto i = 1u; i < arguments.size(); ++i)
+            arg_block.insert(block.getByPosition(arguments[i]));
+        materializeBlockInplace(arg_block);
+        auto out = FormatFactory::instance().getOutputFormat(format_name, buffer, arg_block, context, [&](const Columns &, size_t row)
+        {
+            if constexpr (no_newline)
+            {
+                // replace '\n' with '\0'
+                if (*buffer.position() == '\n')
```

Comment:
I guess it's pointing after the written data not to the last byte.

Reply:
`\n` is correctly replaced as shown in the query test....

Reply:
It's indeed buffer overrun and perfectly captured by Memsan